SafeguardPasswords/test/Unit/TestResponseHandler.query.pq

```diff
@@ -0,0 +1,123 @@
+section TestResponseHandler;
+
+ValidateByStatusCode = Extension.ImportFunction("ValidateByStatusCode", "ResponseHandler.pqm");
+
+TestValidateByStatusCodeRaisesError = () =>
+    let
+        url = "dummy_url",
+        expectedReason = "invalid_request",
+        expectedMessage = "Access denied.",
+        errorCode = 401,
+
+        expectedResponse = [ 
+                    HasError = true, 
+                    Error = [ 
+                                Reason = "Authentication Error", 
+                                Message = "Error 4010: The username or password you have specified is invalid.", 
+                                Detail = [
+                                            Cause = [
+                                                        fakeError = "invalid_request",
+                                                        error_description = "Access denied."
+                                                    ],
+                                                    RequestUrl = "dummy_url", 
+                                                    RelativePath = "fake/relative/path", 
+                                                    ManuallyHandled = true, 
+                                                    ErrorCode = 4010 
+                                            ],
+                                Message.Format = null, 
+                                Message.Parameters = null 
+                            ]
+                ],
+        fakeResponse = FakeRawResponse([    fakeError = expectedReason, 
+                                            error_description = expectedMessage 
+                                        ], 
+                                        [   Response.Status = errorCode, 
+                                            RelativePath = "fake/relative/path", 
+                                            RequestUrl=url
+                                        ]),
+
+        json = Json.Document(fakeResponse),
+
+        responseWithMeta = Value.ReplaceMetadata(json, Value.Metadata(fakeResponse)),
+
+        response = try ValidateByStatusCode(responseWithMeta),
+
+        facts = {
+            Fact("Error response is correct", expectedResponse, response)
+        }
+    in
+        facts;
+
+TestValidateBy406StatusCodeRaisesError = () =>

```

Comment:
I don't think this test is needed, it's enough to extend the TestCommonRequestErros.query.pq in common code.

Reply:
Maybe I misunderstand the TestcommonRequestErros file. As far as I can tell it just makes sure that the error codes are what we expect them to be (compares the hard-coded list of error codes in the test to the error codes returned from CommonRequestErrors.pqm). 

That is not at all what these tests do. Are you saying they should be moved to TestCommonRequestErros.query.pq?

If so, I'm not sure how that would work either. These tests ensure that SafeguardPasswords.ResponseHandler.ValidateByStatusCode returns the expected response - which unfortunately isn't as straightforward as I'd like because the SPP's RSTS returns an unusual error response for login failure (see comment here https://github.com/OneIdentity/SafeguardPowerBI/pull/72/files#diff-382654921fba102a443de86c68df8826e5a266bb6a2b2cb60c9fe35bb4ee549aR27-R29)

I think that the tests should probably stay where they are given that they test logic unique to the SPP Connector, or they should be removed if they are not valuable. 

If they absolutely must be moved however, let me know, and we can discuss how we should go about doing that to test SPP connector logic from the common project.

Reply:
In TestCommonRequestErros.query.pq, we check if our returned error mapping is right (406 should now be included). 

In TestRespondhandler.query.pq, we test if the response handler is working well for the responses. Happy Path: TestValidateByStatusCodeFallbackIsWorking; Unhappy Path: TestValidateByStatusCodeRaisesError, TestValidateBy406StatusCodeRaisesError. 

Here, I think one test is enough for the unhappy path. I think you can delete the TestValidateByStatusCodeRaisesError test and rename the TestValidateBy406StatusCodeRaisesError test without the number.